apps/editing-toolkit/editing-toolkit-plugin/tutorials/class-wpcom-tutorials.php

```diff
@@ -0,0 +1,253 @@
+			&& isset( $task['title'] )
+			&& is_string( $task['title'] );
+	}
+
+	/**
+	 * Retrieves a Tutorial from the registry.
+	 *
+	 * @param string $id  The ID of the Tutorial to fetch.
+	 * @return null|array Tutorial associative array if found, null if not.
+	 */
+	public function get_tutorial( $id ) {
+		if ( ! isset( $this->registry[ $id ] ) ) {
+			return null;
+		}
+		$tutorial = $this->registry[ $id ];
+
+		// Populating status at runtime so that any invalidations happen within the Options API.
+		$tutorial['tasks'] = $this->get_tasks( $tutorial );
+		return $tutorial;
+	}
+
+	/**
+	 * Retrives all registered Tutorial IDs
+	 *
+	 * @return array list of registered Tutorial IDs.
+	 */
+	public function get_registered_ids() {
+		return array_keys( $this->registry );
+	}
+
+	/**
+	 * Gets registered tasks for a Tutorial, with stored completion status.
+	 *
+	 * @param array $tutorial Tutorial object.
+	 * @return array          Tasks for Tutorial, with task completion status.
+	 */
+	private function get_tasks( $tutorial ) {
+		$tasks_with_statuses = array();
+		$statuses            = $this->get_statuses( $tutorial['id'] );
+
+		foreach ( $tutorial['tasks'] as $task ) {
+			// Default status is pending.
+			$task['status'] = 'pending';
+			if ( isset( $statuses[ $task['id'] ] ) ) {
+				$task['status'] = $statuses[ $task['id'] ];
+			}
+			$tasks_with_statuses[] = $task;
+		}
+
+		return $tasks_with_statuses;
+	}
+
+	/**
+	 * Gets the completion statuses for a specific Tutorial.
+	 *
+	 * @param string $id Tutorial ID.
+	 * @return array     Completion statuses for Tutorial ID.
+	 */
+	public function get_statuses( $id ) {
+		$statuses = $this->load_persisted_statuses();
+		return isset( $statuses[ $id ] ) ? $statuses[ $id ] : array();
+	}
+
+	/**
+	 * Loads all Tutorial statuses from the database.
+	 *
+	 * @return array All Tutorial completion statuses.
+	 */
+	private function load_persisted_statuses() {
+		if ( function_exists( 'get_user_attribute' ) ) {
+			return get_user_attribute( get_current_user_id(), 'wpcom_tutorials_status' );
+		}
+		return (array) get_user_meta( get_current_user_id(), 'wpcom_tutorials_status', true );
+	}
+
+	/**
+	 * Marks a task with a status.
+	 *
+	 * @param string $tutorial_id Tutorial ID.
+	 * @param string $task_id     Task ID.
+	 * @param string $status      Status for the task. Allowed: complete|skipped|pending.
+	 * @return bool               True if status is updated, false otherwise.
+	 */
+	public function mark_task( $tutorial_id, $task_id, $status ) {
+		if ( ! $this->validate_status( $status ) ) {
+			return false;
+		}
+		if ( ! $this->task_exists( $tutorial_id, $task_id ) ) {
+			return false;
+		}
+
+		$statuses = $this->load_persisted_statuses();
+		if ( ! isset( $statuses[ $tutorial_id ] ) ) {
+			$statuses[ $tutorial_id ] = array();
+		}
+		// pending is the natural state with no stored status.
+		if ( 'pending' === $status ) {
+			if ( isset( $statuses[ $tutorial_id ][ $task_id ] ) ) {
```

Comment:
>  it won't matter much since reverting to pending (aka unchecking an already-completed step in the UI) won't be that common

Even if it was common, it seems like the current handling will still work fine for this action. Or am I missing something?

Reply:
It will work, yup!